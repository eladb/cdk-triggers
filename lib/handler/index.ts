import * as AWS from 'aws-sdk';

export async function handler(event: AWSLambda.CloudFormationCustomResourceEvent) {
  console.log({ event });

  if (event.RequestType === 'Delete') {
    console.log('not calling trigger on DELETE');
    return;
  }

  const handlerArn = event.ResourceProperties.HandlerArn;
  if (!handlerArn) {
    throw new Error(`The "HandlerArn" property is required`);
  }

  const lambda = new AWS.Lambda();
  const invokeRequest: AWS.Lambda.InvocationRequest = {
    FunctionName: handlerArn,
  };
  console.log({ invokeRequest });
  const invokeResponse = await lambda.invoke(invokeRequest).promise();
  console.log({ invokeResponse });
}