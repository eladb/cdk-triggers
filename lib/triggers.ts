import { Construct, CustomResource, CustomResourceProvider, CustomResourceProviderRuntime } from "@aws-cdk/core";
import * as lambda from '@aws-cdk/aws-lambda';
import { join } from "path";

export interface AfterCreateProps {
  /**
   * Resources to trigger on. Resources can come from any stack in the app.
   *
   * @default [] Run the trigger at any time during stack deployment.
   */
  readonly resources?: Construct[];

  /**
   * The handler to execute once after all the resources are created.
   */
  readonly handler: lambda.IFunction;
}

export class AfterCreate extends Construct {
  constructor(scope: Construct, id: string, props: AfterCreateProps) {
    super(scope, id);

    const provider = CustomResourceProvider.getOrCreateProvider(this, 'AWSCDK.TriggerCustomResourceProvider', {
      runtime: CustomResourceProviderRuntime.NODEJS_12,
      codeDirectory: join(__dirname, 'handler'),
      policyStatements: [
        {
          Effect: 'Allow',
          Action: ['lambda:InvokeFunction'],
          Resource: ['*'], // TODO?
        }
      ]
    });

    const resource = new CustomResource(this, 'Resource', {
      resourceType: 'Custom::Trigger',
      serviceToken: provider.serviceToken,
      properties: {
        HandlerArn: props.handler.functionArn
      }
    });

    // add a dependency between our resource and the resources we want to invoke
    // after.
    resource.node.addDependency(...props.resources ?? []);
  }
}