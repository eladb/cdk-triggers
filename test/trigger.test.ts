import '@aws-cdk/assert/jest';
import { ResourcePart, SynthUtils } from '@aws-cdk/assert';
import { Stack } from '@aws-cdk/core';
import * as sns from '@aws-cdk/aws-sns';
import * as lambda from '@aws-cdk/aws-lambda';
import * as triggers from '../lib';

test('minimal usage', () => {
  // GIVEN
  const stack = new Stack();
  const triggeringResource = new sns.Topic(stack, 'MyTopic');
  const trigger = new lambda.Function(stack, 'MyTriggerHandler', {
    runtime: lambda.Runtime.NODEJS_12_X,
    code: lambda.Code.fromInline('foobar'),
    handler: 'index.handler'
  });

  // WHEN
  new triggers.AfterCreate(stack, 'MyTrigger', {
    handler: trigger,
    resources: [triggeringResource]
  });

  // THEN
  expect(SynthUtils.synthesize(stack).template).toMatchSnapshot();

  expect(stack).toHaveResource('Custom::Trigger', {
    HandlerArn: {
      "Fn::GetAtt": [
        "MyTriggerHandlerD6B1FF23",
        "Arn"
      ]
    }
  });

  expect(stack).toHaveResource('Custom::Trigger', {
    DependsOn: [ "MyTopic86869434" ],
  }, ResourcePart.CompleteDefinition);
});
