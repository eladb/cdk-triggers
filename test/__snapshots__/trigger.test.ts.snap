// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`minimal usage 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters3c9152cc6494b29fcb0e07db4f664cbe2552770efaecad038ac10c5e83436352ArtifactHash32E1B19A": Object {
      "Description": "Artifact hash for asset \\"3c9152cc6494b29fcb0e07db4f664cbe2552770efaecad038ac10c5e83436352\\"",
      "Type": "String",
    },
    "AssetParameters3c9152cc6494b29fcb0e07db4f664cbe2552770efaecad038ac10c5e83436352S3BucketF699E8DE": Object {
      "Description": "S3 bucket for asset \\"3c9152cc6494b29fcb0e07db4f664cbe2552770efaecad038ac10c5e83436352\\"",
      "Type": "String",
    },
    "AssetParameters3c9152cc6494b29fcb0e07db4f664cbe2552770efaecad038ac10c5e83436352S3VersionKey7DE82419": Object {
      "Description": "S3 key for asset version \\"3c9152cc6494b29fcb0e07db4f664cbe2552770efaecad038ac10c5e83436352\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "AWSCDKTriggerCustomResourceProviderCustomResourceProviderHandler97BECD91": Object {
      "DependsOn": Array [
        "AWSCDKTriggerCustomResourceProviderCustomResourceProviderRoleE18FAF0A",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters3c9152cc6494b29fcb0e07db4f664cbe2552770efaecad038ac10c5e83436352S3BucketF699E8DE",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters3c9152cc6494b29fcb0e07db4f664cbe2552770efaecad038ac10c5e83436352S3VersionKey7DE82419",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters3c9152cc6494b29fcb0e07db4f664cbe2552770efaecad038ac10c5e83436352S3VersionKey7DE82419",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": Object {
          "Fn::GetAtt": Array [
            "AWSCDKTriggerCustomResourceProviderCustomResourceProviderRoleE18FAF0A",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "AWSCDKTriggerCustomResourceProviderCustomResourceProviderRoleE18FAF0A": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "lambda:InvokeFunction",
                  ],
                  "Effect": "Allow",
                  "Resource": Array [
                    "*",
                  ],
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "Inline",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "MyTopic86869434": Object {
      "Type": "AWS::SNS::Topic",
    },
    "MyTriggerB6CCCACE": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "MyTopic86869434",
      ],
      "Properties": Object {
        "HandlerArn": Object {
          "Fn::GetAtt": Array [
            "MyTriggerHandlerD6B1FF23",
            "Arn",
          ],
        },
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "AWSCDKTriggerCustomResourceProviderCustomResourceProviderHandler97BECD91",
            "Arn",
          ],
        },
      },
      "Type": "Custom::Trigger",
      "UpdateReplacePolicy": "Delete",
    },
    "MyTriggerHandlerD6B1FF23": Object {
      "DependsOn": Array [
        "MyTriggerHandlerServiceRoleFC0CFFAB",
      ],
      "Properties": Object {
        "Code": Object {
          "ZipFile": "foobar",
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "MyTriggerHandlerServiceRoleFC0CFFAB",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "MyTriggerHandlerServiceRoleFC0CFFAB": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
